doc/oomd/SONiC OOM daemon.md

```diff
@@ -0,0 +1,255 @@
+ - sonic not enable swap for most device.
+ - When user run some command trigger OOM, SONiC will kernel panic. for example:
+   - Multiple user login to device, some service may create 10+ concurrent sessions login to device.
+   - Some user script/command take too much memory, currently 'show' command will take 70 MB memory.
+ - To fix this issue, we need 2 feature: login session limit and OOM daemon.
+
+# 1 Functional Requirement
+ - When exceed memory high-water mark, terminate user process to free memory and protect device from OOM happen.
+
+ - Can set system memory high-water mark and low-water mark.
+ - Can set memory protect policy:
+    - User/group list, their process can be terminate safely.
+    - Terminate policy: terminate user session or terminate user process.
+
+- All process start by local user will not be terminate by OOMD, unless they add to user/group list.
+- Provide default setting to minimize customer side change.
+  - Customers may have pipelines to initialize device configuration, because this feature add new commands, the pipeline may need update. The default limitation is designed to cover most scenario to minimize the customer side change.
+
+# 2 Configuration and Management Requirements
+## 2.1 SONiC CLI
+ - Manage login session or memory  limit settings
+```
+    # config OOMD enable/disable status
+    config oomd { enable/disable}
+    
+    # config oomd policy
+    config oomd policy { policyname } { add|del } <value>
+```
+ - Show oomd config
+```
+    show oomd policy
+    show oomd status
+```
+
+## 2.2 Config DB
+ - OOMD feature can be enable/disable by config DB
+ - OOMD policy are fully configurable by config DB.
+
+# 3 Design
+ - High-level design diagram:
+
+```mermaid
+graph TD;
+%% SONiC CLI update config DB
+CLI[SONiC CLI] -- update OOMD config --> CONFDB[(Config DB)];
+
+
+%% HostCfgd/OOMD subscribe config DB change
+CONFDB -.-> HOSTCFGD[Hostcfgd];
+CONFDB -. config change event .-> OOMD[OOMD];
+
+%% HostCfgd start/stop OOMD
+HOSTCFGD -- start/stop OOMD --> systemd[systemd];
+systemd --> OOMD;
+
+%% OOMD subscript memory utilzation event from cgroup
+CGROUP(cgroup) -. memory utilzation event .-> OOMD; 
+
+%% OOMD get user info from /etc/passwd
+passwd(/etc/passwd) -. user information .-> OOMD;
+
+%% OOMD terminate user session/procress when memory utilzation reach high-water mark
+OOMD -- terminate signal --> process[process]
+
+
+```
+
+## 3.1 OOMD Implementation
+ - OOMD will be a service managed by systemd.
+ - Get system memory information from /proc/meminfo.
+ - Subscript cgroup memory utilization event from /sys/fs/cgroup/memory/memory.usage_in_bytes.
+ - Get user information from /etc/passwd.
+ - OOMD will send SIGKILL/SIGTERM to kill/terminate process.
+ - OOMD can config to terminate user session, this is because when swap not enabled, Linux system may enters a livelocked state much more quickly and may prevent oomd from responding in a reasonable amount of time. 
```

Comment:
>because [](http://example.com/codeflow?start=54&length=7)

I don't understand the cause and effect relationship in this paragraph.

Reply:
Hua: Add more detail, why config disable session.

Reply:
Fixed, add detail in functional requirement part.